creativecommons/cccatalog-api
# Add mechanism based on a boolean mask to keep results consistent

Fixes #314 

## Problem
The problem described in #314 is that when `filter_dead=True` (the default) the amount of results returned might not be equal to the requested `pagesize`. It happens because the result set is filtered to remove broken links after being fetched from ElasticSearch.

In order to return the right amount of results, we could fetch more results from ES in advance (`pagesize * 2`) or when needed until the result set is big enough, but without any way to know where to start the slice of results for each page, it will eventually lead to incorrect pagination where consecutive pages might contain duplicate results.

**Example:**

```python
# 0 → Result with broken link
# 1 → Result with working link
	
results = [1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0]
pagesize = 5

# First page
# Let's say we fetch pagesize * 2 , we would fetch from 0 to 10
# Which contains 5 good results, so we can return those
result_set = [1, 1, 0, 0, 0, 1, 1, 1, 0, 0]

# Second page
# Here we would fetch from 5 to 15
# Which again contains 7 good results, so we could return 5 of those
result_set = [1, 1, 1, 0, 0, 0, 1, 1, 1, 1]

# The problem is the first 5 results of this result_set are the same as 
# the last 5 of the previous one. We would be returning 3 results
# already returned in the previous page!
```


## Proposed solution
A possible solution is to cache a boolean mask of the results already fetched for a certain query.

To uniquely identify each search query, we can serialize elasticsearch_dsl's Search object to a dictionary, strip out the `from` and `size` parameters and use [`DeepHash`](https://deepdiff.readthedocs.io/en/latest/#deep-hash) from [`deepdiff`](https://pypi.org/project/deepdiff/) to hash the dictionary in a deterministic way. We can then use it as our redis key for the mask.

Then when requested for any given page, if we already have a mask for the search query, we can sum it cumulatively and determine at which index the first value of the page appears and start our slice of the results from there.

```python
mask = [1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0]
summed_mask = [1, 2, 2, 2, 2, 3, 4, 5, 5, 5, 5, 6, 7, 8, 9, 10, 10, 10, 10, 10]

# Page 2 requested
# We know that page 2 should start with result nr 6 and end with result nr 10
# So we can use the summed mask to determine where the slice should start
summed_mask.index(6)
```

#### Algorithm
Here's the algorithm used to determine the `start ` and `end` indexes for the result slice.

```python
query_hash = get_query_hash(s)
query_mask = get_query_mask(query_hash)

if not query_mask:
    start = 0
    end = ceil(page_size * page / (1 - DEAD_LINK_RATIO))
elif page_size * (page - 1) > sum(query_mask):
    start = len(query_mask)
    end = ceil(page_size * page / (1 - DEAD_LINK_RATIO))
else:
    accu_query_mask = list(accumulate(query_mask))
    start = 0
    if page > 1:
        try:
            start = accu_query_mask.index(page_size * (page - 1) + 1)
        except ValueError:
            start = accu_query_mask.index(page_size * (page - 1)) + 1
    if page_size * page > sum(query_mask):
        end = ceil(page_size * page / (1 - DEAD_LINK_RATIO))
    else:
        end = accu_query_mask.index(page_size * page) + 1
```

## How to test locally
> Note: The following tests are meant to run on a *nix OS with `curl` and `jq` installed.

With all services running (`docker-compose up -d`), we can switch between `master` and `dead_link_filter` branches (the api server auto-reloads) to see the difference in output of the following commands.

#### Make sure the pagesize is correct when filtering out dead links.
The following command calls the search endpoint with `pagesize=100&filter_dead=True`, then counts the amount of values returned, which should be 100.
```
curl -s "http://localhost:8000/image/search?q=*&pagesize=100&filter_dead=True" | jq '.results | length'
```


#### Make sure that dead links are being correctly filtered out
The following command calls the search endpoint twice with `pagesize=100&filter_dead=True` and `pagesize=100&filter_dead=False`. It then filters the result arrays and passes both to `diff` which should return something like `Files {abc} and {xyz} differ` which means the result arrays have in fact different results.
```
diff -dq <(curl -s "http://localhost:8000/image/search?q=*&pagesize=100&filter_dead=True" | jq '.results') <(curl -s "http://localhost:8000/image/search?q=*&pagesize=100&filter_dead=False" | jq '.results')
```

#### Make sure there's no duplicate results
The following command calls the search endpoint with `pagesize=5&filter_dead=True` from page 1 to 100, filters the _ids_ of the results and stores them in the `result_ids` array. It then sorts and remove duplicate values from the array and counts the number of values left, so it should return 500.
```
declare -a result_ids=(); for i in {1..100}; do result_ids+=(`curl -s "http://localhost:8000/image/search?q=*&pagesize=5&page=${i}&filter_dead=True" | jq '.results[].id'`); done && for i in $result_ids; do echo $i; done | sort -u | wc -l
```
